src/Avalonia.Controls/ItemsControl.cs

```diff
@@ -256,7 +267,14 @@ protected virtual void OnContainersDematerialized(ItemContainerEventArgs e)
                 // when it is removed from the Items collection.
                 if (container?.ContainerControl != container?.Item)
                 {
-                    toRemove.Add(container.ContainerControl);
+                    if (ItemContainerGenerator.ContainerType == null)
+                    {
+                        toRemove.Add((container.ContainerControl as ContentPresenter).Child);
```

Comment:
Again, either use a standard cast here or if we want to handle cases where the item isn't a `ContentPresenter` put in a null check.